### PR TITLE
SR-9 Add Recruiter endpoints

### PR DESCRIPTION
**Pull request by Mendrejk**
Adds all Recruiter endpoints for the MVP.

### CODE REVIEW

**Review by FilipJQ77 — DISMISSED** [review]

File: backend/recruitment-system/src/test/groovy/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/RecruiterServiceSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        when: "The recruiter is added via the service"
+        recruiterService.addRecruiter(alreadyBelongingToTeamId, existingRecruiter)
+
+        then: "An IllegalArgumentException should be thrown and the recruiter should not be added"
+        thrown IllegalArgumentException
+
+        1 * teamRepository.findTeamById(alreadyBelongingToTeamId) >> teamOne
+        1 * recruiterRepository.existsByEmailAndTeamsId(existingRecruiter.email, alreadyBelongingToTeamId) >> true
+        0 * _
+    }
```
Comment (Mario-659) [inline backend/recruitment-system/src/test/groovy/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/RecruiterServiceSpec.groovy:209]: Maybe consider using `thrown` in the order of other statements. In that case it should be like this:
```groovy
1 * teamRepository.findTeamById(alreadyBelongingToTeamId) >> teamOne
1 * recruiterRepository.existsByEmailAndTeamsId(existingRecruiter.email, alreadyBelongingToTeamId) >> true
thrown IllegalArgumentException
0 * _
```

**Review by Mario-659 — CHANGES_REQUESTED** [review]
To fix 👷

File: backend/recruitment-system/src/test/groovy/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/RecruiterServiceSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        def invalidRecruiter = new Recruiter(0, "", [] as Set)
+        def teamId = 2
+
+        when: "The recruiter is added via the service"
+        recruiterService.addRecruiter(teamId, invalidRecruiter)
+
+        then: "An IllegalArgumentException should be thrown and the recruiter should not be added"
+        thrown IllegalArgumentException
+
+        0 * _
```
[the remark this answers is not in the document]
Comment (Mendrejk) [reply backend/recruitment-system/src/test/groovy/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/RecruiterServiceSpec.groovy:177]: Yeah... that's what I meant. I guess strict mocking is not necessary in this test case, or at least, the stories don't suggest it. Your suggestion fits better here. But I think `_` is better than `any()` here.
https://spockframework.org/spock/docs/1.0/interaction_based_testing.html - See Argument Constraints

File: backend/recruitment-system/src/test/groovy/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/RecruiterServiceSpec.groovy (hunk 1 above)
Comment (Mendrejk) [reply backend/recruitment-system/src/test/groovy/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/RecruiterServiceSpec.groovy:209]: Does it change anything in the test logic? I don't think it does but I might be mistaken. For me, just stating that the exception is thrown and separating it from the Mocks looks more readable here.

Comment (Mario-659) [reply backend/recruitment-system/src/test/groovy/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/RecruiterServiceSpec.groovy:209]: No, it seems to not make any difference in test logic. You are right about readability

**Review by Mario-659 — CHANGES_REQUESTED** [review]
> Yeah... that's what I meant. I guess strict mocking is not necessary in this test case, or at least, the stories don't suggest it. Your suggestion fits better here. But I think `_` is better than `any()` here. https://spockframework.org/spock/docs/1.0/interaction_based_testing.html - See Argument Constraints

You are right about `_`. Check your tests because there are still some `0 * _` statements

**Comment by Mendrejk** [comment]
> 

That's true, I've changed all but one `0 * _` to specific method call checks. I feel like the one in the last test makes sense as otherwise I would need to check at least 2 methods for 0 calls (save & delete). I've changed the user story to better signify that no changes should be made.

**Comment by Mario-659** [comment]
Ohh okay, so with `0 * _` test will fail if any method from any mocked object is called (not including methods specified earlier in then block)? I didn't find this statement anywhere. Keep in mind that you can also narrow this down to one object like this:
```groovy
0 * mockedObj._
```

**Comment by Mendrejk** [comment]
> Ohh okay, so with `0 * _` test will fail if any method from any mocked object is called (not including methods specified earlier in then block)? I didn't find this statement anywhere. Keep in mind that you can also narrow this down to one object like this:
> ```groovy
> 0 * mockedObj._
> ```

Yup, it works exactly as you described.
See https://spockframework.org/spock/docs/1.0/interaction_based_testing.html - Strict Mocking

**Review by Bravo555 — APPROVED** [review]

**Review by deamacto — APPROVED** [review]

**Pull request merged**